### PR TITLE
KeyInfo fixes

### PR DESCRIPTION
**Pull request by strehle**
Taken from https://github.com/cloudfoundry/uaa/pull/2208

Problems: 
* Duplicated code ,[ Sonar rated 25% duplicated code ](https://sonarcloud.io/component_measures?metric=duplicated_lines_density&selected=cloudfoundry-identity-parent%3Aserver%2Fsrc%2Fmain%2Fjava%2Forg%2Fcloudfoundry%2Fidentity%2Fuaa%2Foauth%2FKeyInfo.java&view=list&id=cloudfoundry-identity-parent)
* Remove abstract class and combine it into one - remove dublication
* MAC for kty is not RFC complaint https://www.rfc-editor.org/rfc/rfc7518#section-6.1 . To ensure that MAC still works, support both, MAC and oct 
* Add support for EC verification - needed if OIDC Identity Provider is configured with ES??? alg. This might be the case because ES256 is recommended in [OIDC additional standards](https://openid.bitbucket.io/fapi/fapi-2_0-security-profile.html#name-cryptography-and-secrets)

### CODE REVIEW

**Comment by strehle** [comment]
sonar
https://sonarcloud.io/summary/new_code?id=cloudfoundry-identity-parent&pullRequest=2284

before
https://sonarcloud.io/component_measures?id=cloudfoundry-identity-parent&metric=new_duplicated_lines_density&view=list

**Comment by bruce-ricard** [comment]
It looks like you tried splitting this away from another PR you made earlier.

My initial gut reaction is that this PR is still too big. It's very hard to review. I tried to take a peak, there are lots of different changes.

**Comment by strehle** [comment]
> . It's very hard to review. I tried to take a peak, there are lots of different changes.

Ok, so @tack-sap @adrianhoelzl-sap could you please review the change.

The only thing which is new. UAA would accept now ES256 to ES512 algorithms because I activated this in spring-security-jwt.

File: server/src/main/java/org/cloudfoundry/identity/uaa/oauth/KeyInfo.java [hunk 1]
```diff
@@ -19,298 +22,169 @@
...
-        }
-
-        return UriComponentsBuilder.fromHttpUrl(keyUrl).scheme("https").path("/token_keys").build().toUriString();
-    }
-}
-
-class HmacKeyInfo extends KeyInfo {
-    private static final String DEFAULT_HMAC_ALGORITHM = "HMACSHA256";
+public class KeyInfo {
+    private final boolean isAsymetric;
```
Comment (adrianhoelzl-sap) [inline server/src/main/java/org/cloudfoundry/identity/uaa/oauth/KeyInfo.java]: Small typo here, this should be `isAsymmetric`

File: server/src/main/java/org/cloudfoundry/identity/uaa/oauth/KeyInfo.java [hunk 2]
```diff
@@ -19,298 +22,169 @@
...
+            this.verifierCertificate = Optional.empty();
+            this.type = MAC;
+        }
     }
-
-    @Override
     public void verify() {
-
+        // not in use
     }
```
Comment (adrianhoelzl-sap) [inline server/src/main/java/org/cloudfoundry/identity/uaa/oauth/KeyInfo.java]: Can we get rid of the method?

File: server/src/main/java/org/cloudfoundry/identity/uaa/oauth/KeyInfo.java [hunk 3]
```diff
@@ -19,298 +22,169 @@
...
         }
 
-        return result;
+        return UriComponentsBuilder.fromHttpUrl(keyUrl).scheme("https").path("/token_keys").build().toUriString();
     }
 
-    @Override
-    public String algorithm() {
-        return JwtAlgorithms.sigAlg(verifier.algorithm());
+    private static boolean isAssymetricKey(String key) {
```
Comment (adrianhoelzl-sap) [inline server/src/main/java/org/cloudfoundry/identity/uaa/oauth/KeyInfo.java]: Same here, this should be `isAsymmetric`

File: server/src/main/java/org/cloudfoundry/identity/uaa/oauth/KeyInfo.java [hunk 4]
```diff
@@ -19,298 +22,169 @@
...
+        this.isAsymetric = isAssymetricKey(signingKey);
+        String algorithm;
+        if (this.isAsymetric) {
+            String jwtAlg;
+            KeyPair keyPair;
+            try {
+                jwk = JWK.parseFromPEMEncodedObjects(signingKey);
+                jwtAlg = jwk.getKeyType().getValue();
+                if (jwtAlg.startsWith("RSA")) {
+                    algorithm = Optional.ofNullable(sigAlg).map(JwtAlgorithms::sigAlgJava).orElse("SHA256withRSA");
```
Comment (adrianhoelzl-sap) [inline server/src/main/java/org/cloudfoundry/identity/uaa/oauth/KeyInfo.java]: can we extract the fallback value `"SHA256withRSA"` to a constant?

File: server/src/main/java/org/cloudfoundry/identity/uaa/oauth/KeyInfo.java [hunk 5]
```diff
@@ -19,298 +22,169 @@
...
+                    keyPair = jwk.toRSAKey().toKeyPair();
+                    PublicKey rsaPublicKey = keyPair.getPublic();
+                    this.signer = new RsaSigner((RSAPrivateKey) keyPair.getPrivate(), algorithm);
+                    this.verifier = new RsaVerifier((RSAPublicKey) rsaPublicKey, algorithm);
+                    this.type = RSA;
+                } else if (jwtAlg.startsWith("EC")) {
+                    algorithm = Optional.ofNullable(sigAlg).map(JwtAlgorithms::sigAlgJava).orElse("SHA256withECDSA");
```
Comment (adrianhoelzl-sap) [inline server/src/main/java/org/cloudfoundry/identity/uaa/oauth/KeyInfo.java]: same here

File: server/src/main/java/org/cloudfoundry/identity/uaa/oauth/KeyInfo.java [hunk 6]
```diff
@@ -19,298 +22,169 @@
...
+                    throw new IllegalArgumentException("Invalid JWK");
+                }
+            } catch (JOSEException e) {
+                throw new IllegalArgumentException(e);
+            }
+            this.verifierCertificate = Optional.ofNullable(signingCert);
+            this.verifierKey = JsonWebKey.pemEncodePublicKey(keyPair.getPublic()).orElse(null);
+        } else {
+            jwk = new OctetSequenceKey.Builder(signingKey.getBytes()).build();
+            algorithm = Optional.ofNullable(sigAlg).map(JwtAlgorithms::sigAlgJava).orElse("HMACSHA256");
```
Comment (adrianhoelzl-sap) [inline server/src/main/java/org/cloudfoundry/identity/uaa/oauth/KeyInfo.java]: and here

File: server/src/main/java/org/cloudfoundry/identity/uaa/oauth/KeyInfo.java [hunk 7]
```diff
@@ -19,298 +22,169 @@
...
+            }
+            return result;
+        } else {
+            Map<String, Object> result = new HashMap<>();
+            result.put(HeaderParameterNames.ALGORITHM, this.algorithm());
+            result.put(JsonWebKey.PUBLIC_KEY_VALUE, this.verifierKey);
+            //new values per OpenID and JWK spec
+            result.put(JWKParameterNames.PUBLIC_KEY_USE, JsonWebKey.KeyUse.sig.name());
+            result.put(HeaderParameterNames.KEY_ID, this.keyId);
+            result.put(JWKParameterNames.KEY_TYPE, type.name());
```
Comment (adrianhoelzl-sap) [inline server/src/main/java/org/cloudfoundry/identity/uaa/oauth/KeyInfo.java]: Would it make sense to move these lines (except for the one setting `PUBLIC_KEY_VALUE`) above the `if`-statement? Then we could also get rid of them in the first lines of the other branch.

**Review by adrianhoelzl-sap — APPROVED** [review]
Had a look together with @tack-sap, looks good to us

**Pull request merged**